File: python/src/aoc/year2015/day4.py
import hashlib

from aoc.util import load_input


def check(s, zeros):
    return hashlib.md5(s.encode()).hexdigest().startswith(zeros * "0")


def search(base, zeros):
    """
    >>> search('abcdef', 5)
    609043
    >>> search('pqrstuv', 5)
    1048970
    """
    i = 0
    while True:
        if check(base + str(i), zeros):
            return i
        i += 1


def part1(lines):
    return search(lines[0], 5)


def part2(lines):
    return search(lines[0], 6)


if __name__ == "__main__":
    data = load_input(__file__, 2015, "4")
    print(part1(data))
    print(part2(data))

File: rust/src/aoc/year2015/day4.rs
use md5;

pub fn yd() -> (usize, usize) {
    (2015, 4)
}

fn solve(data: &str, target: &str) -> usize {
    let mut i = 0;
    loop {
        let digest = md5::compute(data.to_owned() + &i.to_string());
        if format!("{digest:x}").starts_with(target) {
            return i;
        }
        i += 1;
    }
}

pub fn part1(data: &str) -> usize {
    solve(&data, "00000")
}

pub fn part2(data: &str) -> usize {
    solve(&data, "000000")
}
